operator/src/monitoring.rs

```diff
@@ -0,0 +1,591 @@
+    pub client: reqwest::Client,
+    pub headers: HeaderMap,
+}
+
+impl NifiRestClient {
+    pub fn new(client: reqwest::Client) -> Self {
+        let headers: Vec<(&'static str, &str)> = vec![
+            ("Accept-Encoding", "gzip, deflate, br"),
+            ("Content-Type", "application/json"),
+            ("Accept", "application/json"),
+        ];
+
+        let mut header_map = HeaderMap::new();
+
+        for (key, value) in headers {
+            match HeaderValue::from_str(value) {
+                Ok(val) => {
+                    header_map.insert(key, val);
+                }
+                Err(err) => {
+                    debug!(
+                        "Invalid header item [{} -> {}] for monitoring requests: {}",
+                        key,
+                        value,
+                        err.to_string()
+                    );
+                }
+            }
+        }
+
+        NifiRestClient {
+            client,
+            headers: header_map,
+        }
+    }
+
+    /// Find the "StackablePrometheusReportingTask" in the NiFi cluster.
+    ///
+    /// # Arguments
+    /// * `monitoring_info` - Pod monitoring info including container ports, uid etc.
+    /// * `nifi_version` - The NiFi version currently used by the operator.
+    ///
+    pub async fn find_reporting_task(
+        &self,
+        monitoring_info: &[PodMonitoringInfo],
+        nifi_version: &str,
+    ) -> Result<Option<ReportingTask>, NifiMonitoringError> {
+        let tasks = self.list_reporting_tasks(monitoring_info).await?;
+        for task in tasks {
+            if let Some(ReportingTaskComponent {
+                typ,
+                name,
+                bundle:
+                    ReportingTaskComponentBundle {
+                        group,
+                        artifact,
+                        version,
+                    },
+                ..
+            }) = &task.component
+            {
+                if typ != PROMETHEUS_REPORTING_TASK_TYPE && name != &build_task_name() {
+                    continue;
+                }
+
+                if group != PROMETHEUS_REPORTING_TASK_BUNDLE_GROUP
+                    && artifact != PROMETHEUS_REPORTING_TASK_BUNDLE_ARTIFACT
+                    && version != nifi_version
+                {
+                    continue;
+                }
+
+                // task type and name, bundle group, artifact and version match
+                return Ok(Some(task));
+            }
+        }
+
+        Ok(None)
+    }
+
+    /// List all available ReportingTasks from a NiFi node REST endpoint.
+    /// Will try pod after pod until the first can be queried via REST.
+    ///
+    /// # Arguments
+    /// * `monitoring_info` - Pod monitoring info including container ports, uid etc.
+    ///
+    pub async fn list_reporting_tasks(
+        &self,
+        monitoring_info: &[PodMonitoringInfo],
+    ) -> Result<Vec<ReportingTask>, NifiMonitoringError> {
```

Comment:
This function never returns an error (NifiMonitoringError). It could just return Vec<ReportingTask>

Reply:
`            match self
                .client
                .get(&url)
                .send()
                .await?
                .json::<ReportingTasks>()
                .await
            {`
This (`send().await?`) can throw a reqwest::Error which is propagated to the NifiMonitoringError.

Reply:
Yes, you are right, I overlooked that. But this means the function stops there and doesn't try the other pods at all. The documentation suggests otherwise.